### ISSUE TITLE
Typing for `component` appears incorrect

### ISSUE THREAD

**Issue by ceigey** [matthewp/haunted#88]
Given the following example on the readme:
```javascript
const App = component(({ name }) => {
  return html`Hello ${name}!`;
});
```

... the following typing in index.d.ts seem incorrect:
```typescript
export function component(renderer: (el: HTMLElement) => TemplateResult, BaseElement?: Function, options?: {
    useShadowDOM: boolean
}): Function;
```

Specifically, I would expect the arrow function passed to `component` in the example to have the following type:
```typescript
({ name }: { name: string }) => TemplateResult
```

... which will cause a type error because `Property 'name' does not exist on type 'HTMLElement'. ts(2339)`.

Assuming any arbitrary parameters can be defined for a function passed to `component`, a solution may be (off the top of my head):
```typescript
export function component<T>(renderer: (props: T) => TemplateResult, BaseElement?: Function, options?: {
    useShadowDOM: boolean
}): Function;
```

**Comment by jpray** [comment]
Thanks for catching this.  Would you be okay with the following?

```
export function component(renderer: (el: HTMLElement & any) => TemplateResult, BaseElement?: Function, options?: {
    useShadowDOM: boolean
}): Function;
```

**Comment by Mateiadrielrafael** [comment]
Maybe <T extends HTMLElement> and then use the T?

**Comment by jpray** [comment]
@Mateiadrielrafael is there any benefit to doing `renderer: <T extends HTMLElement>(el: T)` instead of `renderer: (el: HTMLElement & any)`?  Thanks...

**Comment by ceigey** [comment]
Hmm, I think I missed something by not reviewing the code, so I'll update my idea of what the type should be.

```typescript
export function component<T>(
    renderer: (this: HTMLElement | ShadowRoot , props: T) => TemplateResult,
    BaseElement?: Function,
    options?: {
      useShadowDOM: boolean
}): Function;
```

The only real change from my first go at it is adding `this: HTMLElement | ShadowRoot` to `renderer`'s type.

**Below's my thought process, in case I've made any mistakes or further misunderstandings:**

----
In the original example, all that function needs to be passed is an object of roughly the following type:

```typescript
interface ExampleArgument {
  name: string
}
```

... so there, the `extends HTMLElement` seems superfluous.

But perhaps more important than that are the following:

1. how does `renderer` use its argument?
2. how does `component` use `renderer`?

So let me know if I've misunderstood how things work in Haunted!

### How does `renderer` use its argument?
So, based on my understanding, `renderer` is any functional component (in the style of React), which takes props as input (an object of generic type as far as `component` would presumably be concerned, and presumably uses those props to create a `TemplateResult`.

We don't really care about the relationship of `ArgumentObject -> TemplateResult` as LitHtml doesn't have any requirements like a `HTMLElement`, and we assume `renderer` is also a pure function.

Thus the type:

```typescript
(props: T) => TemplateResult
```

*Take a generic property object of type T, pass it to the renderer, and it will produce a TemplateResult*

### How does `component` use `renderer`?

So it appears [`src/component.js`](https://github.com/matthewp/haunted/blob/master/src/component.js) takes `renderer` and passes it to `Container`, without doing anything with it. So no hard dependency on `HTMLElement` there. ATM that's lines :18 and :21.

I will add that `Component` extends `BaseElement` which by default is `HTMLElement`.

Then in [`src/container.js`](https://github.com/matthewp/haunted/blob/master/src/core.js), we find `renderer` used in `render()` on line :74.

```javascript
let result = this.args ?
      this.renderer.apply(this.host, this.args) :
      this.renderer.call(this.host, this.host);
```

I think this part is causing some confusion. [`Function.prototype.apply()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply)'s first argument sets the value of `this`, and in TypeScript I believe you type that [as follows](https://www.typescriptlang.org/docs/handbook/functions.html) (see "`this` parameters"):

```typescript
(this: HTMLElement | ShadowRoot, props: T) => TemplateResult
```

*The second argument to `apply` is just an array of would-be arguments, so we can keep the remaining arguments' types the same*.

**Comment by chase-moskal** [comment]
@ceigey — this looks correct to me, but i know typescript more than haunted's internals

so long as these assumptions are also correct
- the renderer function's `this` should be the component's root html element, whether that be a shadow root or a light-dom html element -- makes sense to me
- the renderer function's first arguments is a `props` object -- soudns good to me
- the signature of the `props` object can be passed into each `component` call via typescript generic, ie, `component<{coolprop: boolean}>(renderer)` -- seems like a good idea

i think we gucci :100: 

one more thing i wonder: what does the `component` function return?
- here we have it return a `Function`, however i wonder if it should return a generic HTMLElement?
- `component<PropsType, MyComponentType extends HTMLElement>(...): MyComponentType`

**Comment by matthewp** [comment]
@chase-moskal It should indeed return a custom element type. @jpray had some difficulties getting that to work which is why it's currently Function, but if you can fix it that would be great.

**Comment by alfredosalzillo** [comment]
Hi, think the correct type should be something like this:

```typescript
type ComponentLike = HTMLElement | ShadowRoot;
type Component<T extends ComponentLike = HTMLElement> = new(...args?: any[]) => T;
export function component<P, T extends ComponentLike = HTMLElement>(
  renderer: (this: T, el: P) => any,
  BaseElement?: Component<T>,
  options?: {
    useShadowDOM: boolean,
    shadowRootInit?: {
      mode?: string
      delegatesFocus?: boolean,
    }
  }): Component<T>;
```

The component function accept a renderer function with a this of type T (a ComponentLike, default an HTMLElement), and args of type P, the props.
The BaseElement should be a Class object of T.
The component function return a Class object of  T.

**Comment by matthewp** [comment]
Nice! Mind sending a PR?

**Comment by alfredosalzillo** [comment]
Hi @matthewp,
i'll send a PR, eventually improving the other declared function types too.

**Comment by matthewp** [comment]
Closed by #116
